In [18]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 5GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/heart-disease-prediction-using-logistic-regression/framingham.csv


In [19]:
df=pd.read_csv('/kaggle/input/heart-disease-prediction-using-logistic-regression/framingham.csv')

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4238 entries, 0 to 4237
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   male             4238 non-null   int64  
 1   age              4238 non-null   int64  
 2   education        4133 non-null   float64
 3   currentSmoker    4238 non-null   int64  
 4   cigsPerDay       4209 non-null   float64
 5   BPMeds           4185 non-null   float64
 6   prevalentStroke  4238 non-null   int64  
 7   prevalentHyp     4238 non-null   int64  
 8   diabetes         4238 non-null   int64  
 9   totChol          4188 non-null   float64
 10  sysBP            4238 non-null   float64
 11  diaBP            4238 non-null   float64
 12  BMI              4219 non-null   float64
 13  heartRate        4237 non-null   float64
 14  glucose          3850 non-null   float64
 15  TenYearCHD       4238 non-null   int64  
dtypes: float64(9), int64(7)
memory usage: 529.9 KB


In [21]:
df.describe(include='all')

male          age    education  currentSmoker   cigsPerDay  \
count  4238.000000  4238.000000  4133.000000    4238.000000  4209.000000   
mean      0.429212    49.584946     1.978950       0.494101     9.003089   
std       0.495022     8.572160     1.019791       0.500024    11.920094   
min       0.000000    32.000000     1.000000       0.000000     0.000000   
25%       0.000000    42.000000     1.000000       0.000000     0.000000   
50%       0.000000    49.000000     2.000000       0.000000     0.000000   
75%       1.000000    56.000000     3.000000       1.000000    20.000000   
max       1.000000    70.000000     4.000000       1.000000    70.000000   

            BPMeds  prevalentStroke  prevalentHyp     diabetes      totChol  \
count  4185.000000      4238.000000   4238.000000  4238.000000  4188.000000   
mean      0.029630         0.005899      0.310524     0.025720   236.721585   
std       0.169584         0.076587      0.462763     0.158316    44.590334   
min       0.000000         0.000000      0.000000     0.000000   107.000000   
25%       0.000000         0.000000      0.000000     0.000000   206.000000   
50%       0.000000         0.000000      0.000000     0.000000   234.000000   
75%       0.000000         0.000000      1.000000     0.000000   263.000000   
max       1.000000         1.000000      1.000000     1.000000   696.000000   

             sysBP        diaBP          BMI    heartRate      glucose  \
count  4238.000000  4238.000000  4219.000000  4237.000000  3850.000000   
mean    132.352407    82.893464    25.802008    75.878924    81.966753   
std      22.038097    11.910850     4.080111    12.026596    23.959998   
min      83.500000    48.000000    15.540000    44.000000    40.000000   
25%     117.000000    75.000000    23.070000    68.000000    71.000000   
50%     128.000000    82.000000    25.400000    75.000000    78.000000   
75%     144.000000    89.875000    28.040000    83.000000    87.000000   
max     295.000000   142.500000    56.800000   143.000000   394.000000   

        TenYearCHD  
count  4238.000000  
mean      0.151958  
std       0.359023  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       0.000000  
max       1.000000

In [22]:
df.isnull().sum()

male                 0
age                  0
education          105
currentSmoker        0
cigsPerDay          29
BPMeds              53
prevalentStroke      0
prevalentHyp         0
diabetes             0
totChol             50
sysBP                0
diaBP                0
BMI                 19
heartRate            1
glucose            388
TenYearCHD           0
dtype: int64

We notice that there are nulls in columns. Lets replace them with mean, median and mode of a particular variable.

The way to input data in NULL Value:

Delete data with complete.cases () if the amount of NULL Value data is not too much

Replace the data with mean () if the data is numeric and there is no outlier

Replace the data with median () if the data is numeric and there is outlier

Replace data with mode(modus) if the data is factorial orstring and the distribution of variations in the data varies. The table () function can help with the mode(modus) of data

For simplicity we will drop all null containing rows. The result is a dataset reduced by 12%


In [28]:
data=df#.drop(['education'],axis=1)
data.dropna(axis=0,inplace=True)

data.describe(include='all')

male          age    education  currentSmoker   cigsPerDay  \
count  3656.000000  3656.000000  3656.000000    3656.000000  3656.000000   
mean      0.443654    49.557440     1.979759       0.489059     9.022155   
std       0.496883     8.561133     1.022657       0.499949    11.918869   
min       0.000000    32.000000     1.000000       0.000000     0.000000   
25%       0.000000    42.000000     1.000000       0.000000     0.000000   
50%       0.000000    49.000000     2.000000       0.000000     0.000000   
75%       1.000000    56.000000     3.000000       1.000000    20.000000   
max       1.000000    70.000000     4.000000       1.000000    70.000000   

            BPMeds  prevalentStroke  prevalentHyp     diabetes      totChol  \
count  3656.000000      3656.000000   3656.000000  3656.000000  3656.000000   
mean      0.030361         0.005744      0.311543     0.027079   236.873085   
std       0.171602         0.075581      0.463187     0.162335    44.096223   
min       0.000000         0.000000      0.000000     0.000000   113.000000   
25%       0.000000         0.000000      0.000000     0.000000   206.000000   
50%       0.000000         0.000000      0.000000     0.000000   234.000000   
75%       0.000000         0.000000      1.000000     0.000000   263.250000   
max       1.000000         1.000000      1.000000     1.000000   600.000000   

             sysBP        diaBP          BMI    heartRate      glucose  \
count  3656.000000  3656.000000  3656.000000  3656.000000  3656.000000   
mean    132.368025    82.912062    25.784185    75.730580    81.856127   
std      22.092444    11.974825     4.065913    11.982952    23.910128   
min      83.500000    48.000000    15.540000    44.000000    40.000000   
25%     117.000000    75.000000    23.080000    68.000000    71.000000   
50%     128.000000    82.000000    25.380000    75.000000    78.000000   
75%     144.000000    90.000000    28.040000    82.000000    87.000000   
max     295.000000   142.500000    56.800000   143.000000   394.000000   

        TenYearCHD  
count  3656.000000  
mean      0.152352  
std       0.359411  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       0.000000  
max       1.000000

For simplicity we will drop education and all null containing columns. The result is a dataset reduced by 12%


In [29]:
category=['male','currentSmoker','BPMeds','prevalentStroke','prevalentHyp','diabetes','education']
cols=list(data.columns)
cols.remove('TenYearCHD')
numeric=list(set(cols)-set(category))

Lets do some Chisquared test for independance between categorical variables and predictor variables

In [30]:
for i in category:
    print("-------"+i+"-------")
    contingency_table=pd.crosstab(data[i],data["TenYearCHD"])
    #print('contingency_table :-\n',contingency_table)

    Observed_Values = contingency_table.values 
    #print("Observed Values :-\n",Observed_Values)

    #Expected Values
    import scipy.stats
    b=scipy.stats.chi2_contingency(contingency_table)
    Expected_Values = b[3]
    #print("Expected Values :-\n",Expected_Values)

    deg_freedom=b[2]
    print("Degree of Freedom:-",deg_freedom)

    from scipy.stats import chi2
    chi_square=sum([(o-e)**2./e for o,e in zip(Observed_Values,Expected_Values)])
    chi_square_statistic=chi_square[0]+chi_square[1]
    print("chi-square statistic:-",chi_square_statistic)

    p_value=1-chi2.cdf(x=chi_square_statistic,df=deg_freedom)
    print('p-value:',p_value)

    if p_value<=0.05:
        print("There is a relationship")
    else:
        print("There is no relationship ")

-------male-------
Degree of Freedom:- 1
chi-square statistic:- 30.77300519507496
p-value: 2.900447715337151e-08
There is a relationship
-------currentSmoker-------
Degree of Freedom:- 1
chi-square statistic:- 1.3444080980686755
p-value: 0.24625808474142385
There is no relationship 
-------BPMeds-------
Degree of Freedom:- 1
chi-square statistic:- 29.03452113804038
p-value: 7.109993427345529e-08
There is a relationship
-------prevalentStroke-------
Degree of Freedom:- 1
chi-square statistic:- 8.546916048696804
p-value: 0.0034610809460219327
There is a relationship
-------prevalentHyp-------
Degree of Freedom:- 1
chi-square statistic:- 120.51173022673483
p-value: 0.0
There is a relationship
-------diabetes-------
Degree of Freedom:- 1
chi-square statistic:- 31.89157153229049
p-value: 1.6302300176462836e-08
There is a relationship
-------education-------
Degree of Freedom:- 3
chi-square statistic:- 31.062637414123984
p-value: 8.246210529971876e-07
There is a relationship


From the above results. All categorical variables except currSmoker has a significant relationship with Heart Disease

In [31]:
from scipy import stats
for i in numeric:
    print("------"+i+"--------")
    df_anova = data[["TenYearCHD",i]]

    grps = pd.unique(df_anova["TenYearCHD"].values)
    d_data = {grp:df_anova[i][df_anova["TenYearCHD"] == grp] for grp in grps}

    F, p = stats.f_oneway(d_data[0], d_data[1])
    print("p-value for significance is: ", round(p,4))
    if p<0.05:
        print("reject null hypothesis")
    else:
        print("accept null hypothesis")

------age--------
p-value for significance is:  0.0
reject null hypothesis
------cigsPerDay--------
p-value for significance is:  0.0016
reject null hypothesis
------diaBP--------
p-value for significance is:  0.0
reject null hypothesis
------totChol--------
p-value for significance is:  0.0
reject null hypothesis
------sysBP--------
p-value for significance is:  0.0
reject null hypothesis
------BMI--------
p-value for significance is:  0.0
reject null hypothesis
------heartRate--------
p-value for significance is:  0.2147
accept null hypothesis
------glucose--------
p-value for significance is:  0.0
reject null hypothesis


Heart rate shows no relation with Heart disease.
Lets see the variation inflation factor for Numeric data

Lets take two approaches here.A model with the above filtered variables. One with feature selection by backward elimination.

Case 1 (backward elimination)

We feed the dataset without scrutinizing the independant variable and apply backward elimination using p-value to come with a set of variables for which p values are below (0.05)

In [32]:
from statsmodels.tools import add_constant
from statsmodels.discrete.discrete_model import Logit
x=data[cols]
x = add_constant(x)
y=data.TenYearCHD
model=Logit(y,x)
result=model.fit()
result.summary()

Optimization terminated successfully.
         Current function value: 0.376668
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:             TenYearCHD   No. Observations:                 3656
Model:                          Logit   Df Residuals:                     3640
Method:                           MLE   Df Model:                           15
Date:                Sun, 17 May 2020   Pseudo R-squ.:                  0.1174
Time:                        12:54:39   Log-Likelihood:                -1377.1
converged:                       True   LL-Null:                       -1560.3
Covariance Type:            nonrobust   LLR p-value:                 8.027e-69
===================================================================================
                      coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------
const              -8.3222      0.715    -11.632      0.000      -9.725      -6.920
male                0.5551      0.109      5.090      0.000       0.341       0.769
age                 0.0635      0.007      9.499      0.000       0.050       0.077
education          -0.0475      0.049     -0.962      0.336      -0.144       0.049
currentSmoker       0.0709      0.157      0.452      0.651      -0.236       0.378
cigsPerDay          0.0179      0.006      2.874      0.004       0.006       0.030
BPMeds              0.1623      0.234      0.692      0.489      -0.297       0.621
prevalentStroke     0.6935      0.490      1.417      0.157      -0.266       1.653
prevalentHyp        0.2346      0.138      1.700      0.089      -0.036       0.505
diabetes            0.0395      0.315      0.125      0.900      -0.579       0.658
totChol             0.0023      0.001      2.062      0.039       0.000       0.005
sysBP               0.0154      0.004      4.043      0.000       0.008       0.023
diaBP              -0.0041      0.006     -0.642      0.521      -0.017       0.008
BMI                 0.0066      0.013      0.518      0.605      -0.018       0.032
heartRate          -0.0032      0.004     -0.772      0.440      -0.012       0.005
glucose             0.0071      0.002      3.189      0.001       0.003       0.012
===================================================================================
"""

The message box shows it took 10 iterations to fit the best possible model and also shows current function value (objective function). This value is helpful when after certain number of iterations the model wont learn so the objective function remains the same

Maximum Likelihood estimation: To find an optimal way to fit a distribution on the data. the optimal place at which the distribution must be placed to define the data well. Bigger the likelihood higher the probability that our model is correct.

Log likely hood- its almost negative always. Bigger the better
LL null- log likely hood null	- log likely hood when there is no independant variable. Aka. useless model

The comparison between these two metrics gives 
LLR p-value- show how statistically significant the model is compared to LL-NULL

Pseudo R squared(McFadden’s R-squared) - somewhere between 0.2 and 0.4 good. Useful only to comapare different variation of same model

In [33]:
cols=data.columns[:-1]
print(cols)

Index(['male', 'age', 'education', 'currentSmoker', 'cigsPerDay', 'BPMeds',
       'prevalentStroke', 'prevalentHyp', 'diabetes', 'totChol', 'sysBP',
       'diaBP', 'BMI', 'heartRate', 'glucose'],
      dtype='object')


In [34]:

def back_feature_elem (data_frame,dep_var,col_list):
    while len(col_list)>0 :
        x=data[col_list]
        x = add_constant(x)
        y=data.TenYearCHD
        model=Logit(y,x)
        result=model.fit(disp=0)
        largest_pvalue=round(result.pvalues,3).nlargest(1)
        if largest_pvalue[0]<(0.05):
            return result
            break
        else:
            col_list=col_list.drop(largest_pvalue.index)
    
result=back_feature_elem(data,data.TenYearCHD,cols)
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:             TenYearCHD   No. Observations:                 3656
Model:                          Logit   Df Residuals:                     3649
Method:                           MLE   Df Model:                            6
Date:                Sun, 17 May 2020   Pseudo R-squ.:                  0.1147
Time:                        12:54:39   Log-Likelihood:                -1381.2
converged:                       True   LL-Null:                       -1560.3
Covariance Type:            nonrobust   LLR p-value:                 2.885e-74
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -9.1298      0.476    -19.199      0.000     -10.062      -8.198
male           0.5614      0.107      5.255      0.000       0.352       0.771
age            0.0659      0.006     10.254      0.000       0.053       0.078
cigsPerDay     0.0192      0.004      4.604      0.000       0.011       0.027
totChol        0.0023      0.001      2.024      0.043    7.16e-05       0.004
sysBP          0.0175      0.002      8.159      0.000       0.013       0.022
glucose        0.0073      0.002      4.342      0.000       0.004       0.011
==============================================================================
"""

Feature Selection: Backward elemination (P-value approach)
Takes in the dataframe, the dependent variable and a list of column names, 
runs the regression repeatedly eliminating feature with the highest
P-value above alpha one at a time and returns the regression summary 
with all p-values below alpha

Lets interpret the coefficients of the variables.
The exponential off the coeffcients give the odds ratio.

In [42]:
print(np.exp(result.params))

const         0.000108
male          1.753206
age           1.068116
cigsPerDay    1.019412
totChol       1.002275
sysBP         1.017689
glucose       1.007307
dtype: float64


the values mean that
for males there is a 75% (1.75-1 =0.75) higher chances of heart disease than for females

similarly,
for 1 year increase in age there is 6.8% higher chance of getting heart disease.

and so on..

Lets do some predictions

In [46]:
#lets get the confusion matrix based on statsmodel

conf_sm=pd.DataFrame(result.pred_table())
conf_sm.columns=['predicted 0','predicted 1']
conf_sm=conf_sm.rename(index={0:'Actual 0',1:'Actual 1'})
conf_sm

predicted 0  predicted 1
Actual 0       3076.0         23.0
Actual 1        515.0         42.0

In [52]:
#train split

import sklearn
new_features=data[['age','male','cigsPerDay','totChol','sysBP','glucose','TenYearCHD']]
x=new_features.iloc[:,:-1]
y=new_features.iloc[:,-1]
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=.20,random_state=5)

from sklearn.linear_model import LogisticRegression
logreg=LogisticRegression()
logreg.fit(x_train,y_train)
y_pred=logreg.predict(x_test)


from sklearn import metrics
confusion_matrix = metrics.confusion_matrix(y_test,y_pred)

conf=pd.DataFrame(confusion_matrix)
conf.columns=['predicted 0','predicted 1']
conf=conf.rename(index={0:'Actual 0',1:'Actual 1'})
print(conf)

print("model accuracy:")
print(sklearn.metrics.accuracy_score(y_test,y_pred))

          predicted 0  predicted 1
Actual 0          620            3
Actual 1           97           12
model accuracy:
0.8633879781420765


The model predicts the correct output 86.3% of the time


Case 2-

Lets try for model with feature selection done manually. Based on our hypothesis testing. We rejected currentSmoker and heart rate to be of significant predictor.

In [56]:
x=data.drop(['TenYearCHD','currentSmoker','heartRate'],axis=1)
x=add_constant(x)
y=data.TenYearCHD
model=Logit(y,x)
result=model.fit(disp=0)
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:             TenYearCHD   No. Observations:                 3656
Model:                          Logit   Df Residuals:                     3642
Method:                           MLE   Df Model:                           13
Date:                Sun, 17 May 2020   Pseudo R-squ.:                  0.1171
Time:                        14:27:58   Log-Likelihood:                -1377.5
converged:                       True   LL-Null:                       -1560.3
Covariance Type:            nonrobust   LLR p-value:                 4.151e-70
===================================================================================
                      coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------
const              -8.4739      0.660    -12.834      0.000      -9.768      -7.180
male                0.5649      0.108      5.218      0.000       0.353       0.777
age                 0.0636      0.007      9.563      0.000       0.051       0.077
education          -0.0460      0.049     -0.931      0.352      -0.143       0.051
cigsPerDay          0.0196      0.004      4.664      0.000       0.011       0.028
BPMeds              0.1701      0.234      0.727      0.467      -0.289       0.629
prevalentStroke     0.7015      0.489      1.435      0.151      -0.257       1.660
prevalentHyp        0.2286      0.138      1.659      0.097      -0.041       0.499
diabetes            0.0399      0.315      0.127      0.899      -0.577       0.657
totChol             0.0023      0.001      2.013      0.044    6.01e-05       0.004
sysBP               0.0153      0.004      4.016      0.000       0.008       0.023
diaBP              -0.0044      0.006     -0.685      0.494      -0.017       0.008
BMI                 0.0058      0.013      0.460      0.645      -0.019       0.031
glucose             0.0070      0.002      3.141      0.002       0.003       0.011
===================================================================================
"""

Removing all insignificant variables

In [57]:
x=data.drop(['TenYearCHD','currentSmoker','heartRate','glucose','prevalentStroke','education','BPMeds','prevalentHyp','diabetes','diaBP','BMI'],axis=1)
x=add_constant(x)
y=data.TenYearCHD
model=Logit(y,x)
result=model.fit(disp=0)
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:             TenYearCHD   No. Observations:                 3656
Model:                          Logit   Df Residuals:                     3650
Method:                           MLE   Df Model:                            5
Date:                Sun, 17 May 2020   Pseudo R-squ.:                  0.1087
Time:                        14:30:19   Log-Likelihood:                -1390.7
converged:                       True   LL-Null:                       -1560.3
Covariance Type:            nonrobust   LLR p-value:                 3.803e-71
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -8.6872      0.459    -18.909      0.000      -9.588      -7.787
male           0.5690      0.106      5.344      0.000       0.360       0.778
age            0.0670      0.006     10.464      0.000       0.054       0.079
cigsPerDay     0.0184      0.004      4.434      0.000       0.010       0.027
totChol        0.0023      0.001      2.073      0.038       0.000       0.005
sysBP          0.0183      0.002      8.551      0.000       0.014       0.022
==============================================================================
"""

In [61]:
#train split

import sklearn
new_features=data[['age','male','cigsPerDay','totChol','sysBP','TenYearCHD']]
x=new_features.iloc[:,:-1]
y=new_features.iloc[:,-1]
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=.20,random_state=5)
#print(x_test)
from sklearn.linear_model import LogisticRegression
logreg=LogisticRegression()
logreg.fit(x_train,y_train)
y_pred=logreg.predict(x_test)

from sklearn import metrics
confusion_matrix = metrics.confusion_matrix(y_test,y_pred)

conf=pd.DataFrame(confusion_matrix)
conf.columns=['predicted 0','predicted 1']
conf=conf.rename(index={0:'Actual 0',1:'Actual 1'})
print(conf)

print("model accuracy:")
print(sklearn.metrics.accuracy_score(y_test,y_pred))

[-8.93217051]
          predicted 0  predicted 1
Actual 0          622            1
Actual 1           99           10
model accuracy:
0.8633879781420765


There is only slight difference in accuracy between the two models.